internal/source/git.go

```diff
@@ -1,146 +1,92 @@
+	if err != nil {
+		return nil, fmt.Errorf("bundle unpack git clone error: %w", err)
+	}
+	wt, err := repo.Worktree()
+	if err != nil {
+		return nil, fmt.Errorf("bundle unpack error: %w", err)
+	}
 
-		if len(pod.Spec.Containers) != 1 {
-			pod.Spec.Containers = make([]corev1.Container, 1)
+	// Checkout commit
+	if gitsource.Ref.Commit != "" {
+		commitHash := plumbing.NewHash(gitsource.Ref.Commit)
+		if err := wt.Reset(&git.ResetOptions{
+			Commit: commitHash,
+			Mode:   git.HardReset,
+		}); err != nil {
+			return nil, fmt.Errorf("checkout commit %q: %w", commitHash.String(), err)
 		}
+	}
 
-		pod.Spec.Containers[0].Name = gitBundleUnpackContainerName
-		pod.Spec.Containers[0].Image = g.UnpackImage
-		pod.Spec.Containers[0].ImagePullPolicy = corev1.PullIfNotPresent
-		pod.Spec.Containers[0].Command = []string{"/unpack", "--bundle-dir", "/bundle"}
-		pod.Spec.Containers[0].VolumeMounts = []corev1.VolumeMount{{Name: "bundle", MountPath: "/bundle"}}
+	var bundleFS fs.FS = &billyFS{wt.Filesystem}
 
-		pod.Spec.Volumes = []corev1.Volume{
-			{Name: "bundle", VolumeSource: corev1.VolumeSource{EmptyDir: &corev1.EmptyDirVolumeSource{}}},
+	// Subdirectory
+	if gitsource.Directory != "" {
+		directory := filepath.Clean(gitsource.Directory)
+		if directory[:3] == "../" || directory[0] == '/' {
+			return nil, fmt.Errorf("get subdirectory %q for repository %q: %s", gitsource.Directory, gitsource.Repository, "directory can not start with '../' or '/'")
 		}
-		return nil
-	})
-}
-
-func pendingGitPodResult(pod *corev1.Pod) *Result {
-	var messages []string
-	for _, cStatus := range append(pod.Status.InitContainerStatuses, pod.Status.ContainerStatuses...) {
-		if waiting := cStatus.State.Waiting; waiting != nil {
-			if waiting.Reason == "ErrImagePull" || waiting.Reason == "ImagePullBackOff" {
-				messages = append(messages, waiting.Message)
-			}
+		sub, err := wt.Filesystem.Chroot(filepath.Clean(directory))
+		if err != nil {
+			return nil, fmt.Errorf("get subdirectory %q for repository %q: %w", gitsource.Directory, gitsource.Repository, err)
 		}
+		bundleFS = &billyFS{sub}
 	}
-	return &Result{State: StatePending, Message: strings.Join(messages, "; ")}
-}
-
-func (g *Git) failedPodResult(ctx context.Context, pod *corev1.Pod) error {
-	logs, err := g.getPodLogs(ctx, pod)
-	if err != nil {
-		return fmt.Errorf("unpack failed: failed to retrieve failed pod logs: %w", err)
-	}
-	_ = g.Client.Delete(ctx, pod)
-	return fmt.Errorf("unpack failed: %v", string(logs))
-}
 
-func (g *Git) succeededPodResult(ctx context.Context, bundle *rukpakv1alpha1.Bundle, pod *corev1.Pod) (*Result, error) {
-	bundleFS, err := g.getBundleContents(ctx, pod)
-	if err != nil {
-		return nil, fmt.Errorf("get bundle contents: %w", err)
-	}
 	resolvedSource := &rukpakv1alpha1.BundleSource{
 		Type: rukpakv1alpha1.SourceTypeGit,
 		// TODO: improve git source implementation to return result with commit hash.
```

Comment:
Do we have access to the git commit hash now that we're doing this all with a git library?

I'm curious if we can "resolve" the `bundle.Spec.Source.Git` struct by translating it to one that uses `commit` instead of tag or branch.

Reply:
It seems possible to resolve it to the hash.  `func (*Repository) ResolveRevision (rev plumbing.Revision) (*plumbing.Hash, error)` of `github.com/go-git/go-git/v5/Repository` can translate the tag and branch to the hash then the hash can be used to reset the repo to the commit.

Reply:
I don't think we need to actually reset the repo (unless we have to). I'm just asking if its possible to lookup the hash of the current worktree and then use that in the `status.resolvedSource`.

Reply:
This isn't a blocker, btw. I think we can do this as a follow-up.

Reply:
> can we use ResolveRevision to populate the bundle.status.resolvedSource?

yes, we can use the hash for all types.